### PR TITLE
Rest api

### PR DESCRIPTION
**Pull request by shahrukh-ijaz**
Rest API

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,6 +39,44 @@ You can run the test by using just a command
 
 `./manage.py test`<br>
 
+
+## REST API
+```python\
+# Login [POST]
+http://127.0.0.1:8001/api/login?username=shahrukh-ijaz&password=shahrukh31
```
Comment (asamolion) [inline README.md]: You are passing the params as URL parameters. That is a horrible practice. Read about GET vs POST and how to properly utilize the POST request.

File: contacts/serializers.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class UserSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = User
+        fields = ('id', 'first_name', 'last_name', 'username', 'email')
+
+
+class UserDetailSerializer(serializers.ModelSerializer):
+    # contacts = ContactSerializer(many=True, required=False)
```
Comment (asamolion) [inline contacts/serializers.py]: Never push commented code to github.

File: contacts/urls.py [hunk 3]
```diff
@@ -1,36 +1,10 @@
...
-    path('update_contact/', views.update_contact, name='update_contact'),
-
-    url(r'^activate/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
-        views.activate, name='activate'),
-
+    path('api/profile', UserProfile.as_view(), name='profile'),
+    path('api/contacts', UserContacts.as_view(), name='contacts'),
+    path('api/logout', Logout.as_view(), name='logout'),
+    path('api/login', login),
+    path('api/sampleapi', sample_api)
```
Comment (asamolion) [inline contacts/urls.py]: Why is the sample API necessary?

File: contacts/views.py [hunk 4]
```diff
@@ -1,230 +1,157 @@
...
+    ContactDetailSerializer
+from contacts.models import User, Contact, Email
+from django.contrib.auth import authenticate
+
+
+class UserProfile(APIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def get(self, request):
+        if 'token' in request.COOKIES:
```
Comment (asamolion) [inline contacts/views.py]: You are checking if the token is is present in the cookies without actually verifying whether the value of the token is actually correct or not. Please rectify this.

File: contacts/views.py [hunk 5]
```diff
@@ -1,230 +1,157 @@
...
+            pk = request.COOKIES.get('user_id')
+            user = User.objects.get(id=pk)
+            payload = request.data
+            serializer = UserSerializer(user, data=payload, partial=True)
+            if serializer.is_valid(raise_exception=True):
+                serializer.save()
+            return Response(serializer.data, status=HTTP_200_OK)
+        else:
+            return Response("PUT REQUEST FAILED", status=HTTP_400_BAD_REQUEST)
```
Comment (asamolion) [inline contacts/views.py]: Don't tell the client exactly what failed. A generic error message is enough.

File: contacts/views.py [hunk 6]
```diff
@@ -1,230 +1,157 @@
...
+    def get_serialized(self, pk, user):
+
+        queryset = Contact.objects.filter(user_id=pk)
+
+        serializer = ContactSerializer(queryset, many=True)
+        data = serializer.data
+        return data
+
+
+class UserContacts(APIView):
```
Comment (asamolion) [inline contacts/views.py]: Separate API views in to an `api.py` file

File: contacts/views.py [hunk 7]
```diff
@@ -1,230 +1,157 @@
...
+    permission_classes = (permissions.AllowAny,)
```
Comment (asamolion) [inline contacts/views.py]: Why is `AllowAny` added as a permission.

File: contacts/views.py [hunk 8]
```diff
@@ -1,230 +1,157 @@
...
+
+    def post(self, request):
+        serializer = ContactSerializer(data=request.data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors)
+
+    def put(self, request):
+        if 'token' in request.COOKIES:
```
Comment (asamolion) [inline contacts/views.py]: Use proper authentication and authorization with the token.

File: contacts/views.py [hunk 9]
```diff
@@ -1,230 +1,157 @@
...
+    password = request.GET.get("password")
+    if username is None or password is None:
+        return Response({'error': 'Please provide both username and password'},
+                        status=HTTP_400_BAD_REQUEST)
+    user = authenticate(username=username, password=password)
+    if not user:
+        return Response({'error': 'Invalid Credentials'},
+                        status=HTTP_404_NOT_FOUND)
+
+    token, _ = Token.objects.get_or_create(user=user)
```
Comment (asamolion) [inline contacts/views.py]: why is the underscore necessary?

**Review by asamolion — CHANGES_REQUESTED** [review]

File: contacts/views.py (hunk 9 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: it gives two things in return and we only concern with the authentication token that's why underscore is needed!

File: README.md (hunk 1 above)
Comment (shahrukh-ijaz) [reply README.md]: done

File: contacts/serializers.py (hunk 2 above)
Comment (shahrukh-ijaz) [reply contacts/serializers.py]: done

File: contacts/urls.py (hunk 3 above)
Comment (shahrukh-ijaz) [reply contacts/urls.py]: removed this is for the testing sorry forget to remove

File: contacts/views.py (hunk 4 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]:
```python
        try:
            if 'token' in request.COOKIES:
                token = Token.objects.get(key=request.COOKIES['token'])
            else:
                token = None
        except Token.DoesNotExist:
            token = None
        if token is not None:
```

Comment (shahrukh-ijaz) [reply contacts/views.py]: done

File: contacts/views.py (hunk 5 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: ok

File: contacts/views.py (hunk 6 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: ok

File: contacts/views.py (hunk 7 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: i give every permission but now only base permission

File: contacts/views.py (hunk 8 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: done

File: contacts/api.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+            if 'token' in request.COOKIES:
+                token = Token.objects.get(key=request.COOKIES['token'])
+            else:
+                token = None
+        except Token.DoesNotExist:
+            token = None
+        if token is not None:
+            pk = request.COOKIES.get('user_id')
+            user = User.objects.get(id=pk)
+            # contacts = self.get_serialized(pk, user)
```
Comment (asamolion) [inline contacts/api.py]: Never push commented code to github.

File: contacts/api.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+            serializer = UserSerializer(user)
+            return Response(serializer.data, status=HTTP_200_OK)
+        else:
+            return Response("Your token expire Login again", status=HTTP_400_BAD_REQUEST)
+
+    def put(self, request):
+        try:
```
Comment (asamolion) [inline contacts/api.py]: This token checking code is being duplicated on many API views. please refactor this to make it cleaner.

File: contacts/api.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+        queryset = Contact.objects.filter(user_id=pk)
+
+        serializer = ContactSerializer(queryset, many=True)
+        data = serializer.data
+        return data
+
+
+class UserContacts(APIView):
+    permission_classes = (permissions.BasePermission,)
```
Comment (asamolion) [inline contacts/api.py]: What does BasePermission do?

File: contacts/api.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            token = None
+        if token is not None:
+            contact = Contact.objects.get(id=request.data['id'])
+            payload = request.data
+            serializer = ContactSerializer(contact, data=payload, partial=True)
+            if serializer.is_valid(raise_exception=True):
+                serializer.save()
+            return Response(serializer.data, status=HTTP_200_OK)
+        else:
+            return Response("BAD REQUEST", status=HTTP_400_BAD_REQUEST)
```
Comment (asamolion) [inline contacts/api.py]: There is a difference between 400 bad request and bad authorization. Please use correct HTTP response codes.

File: contacts/views.py (hunk 7 above)
Comment (asamolion) [reply contacts/views.py]: What does BasePermission do?

File: contacts/api.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+
+class ContactDetails(APIView):
+    permission_classes = (permissions.AllowAny,)
```
Comment (asamolion) [inline contacts/api.py]: Please read about Django permissions and manage all permissions  accordingly.

File: contacts/api.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+        serializer = UserDetailSerializer(user)
+        return Response(serializer.data)
+
+    def get_serialized(self, pk, user):
+
+        queryset = Contact.objects.filter(user_id=pk)
+
+        serializer = ContactSerializer(queryset, many=True)
+        data = serializer.data
```
Comment (asamolion) [inline contacts/api.py]: The data variable is unnecessary.

File: contacts/views.py [hunk 16]
```diff
@@ -6,15 +6,25 @@
 from django.template.loader import render_to_string
 from django.utils.encoding import force_bytes, force_text
 from django.utils.http import urlsafe_base64_encode, urlsafe_base64_decode
+# from requests import Response
```
Comment (asamolion) [inline contacts/views.py]: Never push commented code.

**Review by asamolion — CHANGES_REQUESTED** [review]

File: contacts/api.py (hunk 10 above)
Comment (shahrukh-ijaz) [reply contacts/api.py]: ok

File: contacts/api.py (hunk 11 above)
Comment (shahrukh-ijaz) [reply contacts/api.py]:
```python
def validate_token(request):
    try:
        if 'token' in request.COOKIES:
            token = Token.objects.get(key=request.COOKIES['token'])
            return token
        else:
            return None
    except Token.DoesNotExist:
        return None

```

Comment (shahrukh-ijaz) [reply contacts/api.py]: done

File: contacts/api.py (hunk 13 above)
Comment (shahrukh-ijaz) [reply contacts/api.py]:
```
return Response("UNAUTHORIZED", status=HTTP_401_UNAUTHORIZED)
```

File: contacts/api.py (hunk 15 above)
Comment (shahrukh-ijaz) [reply contacts/api.py]: ok done

File: contacts/views.py (hunk 16 above)
Comment (shahrukh-ijaz) [reply contacts/views.py]: ok

File: contacts/api.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+)
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from contacts.serializers import UserSerializer, ContactSerializer, UserDetailSerializer
+from contacts.models import User, Contact
+from django.contrib.auth import authenticate
+from contacts.permissions import IsUserLogin
+
+
+def validate_token(request):
```
Comment (asamolion) [inline contacts/api.py]: Try writing this as a python function decorator.

File: contacts/api.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+class ContactDetails(APIView):
+
+    def get(self, request, pk):
+        book = Contact.objects.get(id=pk)
+        serializer = ContactSerializer(book)
+        return Response(serializer.data)
+
+
+class UserDetails(APIView):
+
```
Comment (asamolion) [inline contacts/api.py]: Use proper PEP8 formatting. Also write docstrings for these methods.

File: contacts/permissions.py [hunk 19]
```diff
@@ -0,0 +1,23 @@
+from tokenize import Token
+
+from rest_framework.permissions import BasePermission
+
+
+def validate_token(request):
```
Comment (asamolion) [inline contacts/permissions.py]: This code is being duplicated. Please DRY this.

File: contacts/tests/api_test.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+        self.profile = "http://127.0.0.1:8001/api/profile"
+        self.username = "shahrukh-ijaz31"
+        self.password = "shahrukh31"
+        self.email = "shahrukh@arbisoft.com"
+
+    def test_get_profile_details(self):
+        self.user = User.objects.create_user(username=self.username, password=self.password, email=self.email)
+        response = self.client.post(self.url, {"username": self.username, "password": self.password},
+                                    format='multipart')
+        if response.status_code == 200:
```
Comment (asamolion) [inline contacts/tests/api_test.py]: What is this if/else block doing?

File: contacts/tests/api_test.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+            response = self.client.get(self.profile)
+            self.assertEqual(200, response.status_code)
+        else:
+            self.assertFalse(200, response.status_code)
+
+    def test_put_profile_details(self):
```
Comment (asamolion) [inline contacts/tests/api_test.py]: Give a more descriptive name to this test.

File: contacts/tests/api_test.py [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+class ContactsAPIViewTestCase(APITestCase):
+
+    def setUp(self):
+        self.url = "http://127.0.0.1:8001/api/login"
+        self.contacts = "http://127.0.0.1:8001/api/contacts"
+        self.username = "shahrukh-ijaz31"
+        self.password = "shahrukh31"
+        self.email = "shahrukh@arbisoft.com"
+
+    def test_get_contacts(self):
```
Comment (asamolion) [inline contacts/tests/api_test.py:67]: Actually create some contacts for that user and then get and see if they are valid or not.

**Review by asamolion — CHANGES_REQUESTED** [review]

File: contacts/tests/api_test.py (hunk 21 above)
Comment (shahrukh-ijaz) [reply contacts/tests/api_test.py]: ok

File: contacts/api.py (hunk 18 above)
Comment (shahrukh-ijaz) [reply contacts/api.py]: Done

File: contacts/tests/api_test.py (hunk 20 above)
Comment (shahrukh-ijaz) [reply contacts/tests/api_test.py]: additional check removed it

File: contacts/tests/api_test.py (hunk 22 above)
Comment (shahrukh-ijaz) [reply contacts/tests/api_test.py:67]: done

**Review by asamolion — APPROVED** [review]

**Review by asamolion — APPROVED** [review]

**Pull request merged**